### PR TITLE
 Adapt SABnzbd for Generic Service Support

### PR DESCRIPTION
**Pull request by Safihre**
@ymartin59 I need your help. Very much a WIP, but I am stuck at the start already.
With the changes I have now the compile wouldn't even start, it says:
```
../../mk/spksrc.service.mk:59: *** Set STARTABLE=no or provide either SERVICE_COMMAND or specific SSS_SCRIPT.  Stop.
```
But that variable is clearly there.

_For now based on your branch, but will later rebase to just be SABnzbd._
### Checklist
- [x] Build rule `all-supported` completed successfully
- [x] Package upgrade completed successfully
- [x] New installation of package completed successfully

### CODE REVIEW

**Comment by Safihre** [comment]
@ymartin59 I am making some more progress already.
Found out that `sabnzbd` needs to compile `python` because the Python-wheel `sabyenc` needs `Pyconfig.h`. 
Although I would expect that a workaround for that should be possible.

**Comment by Safihre** [comment]
@ymartin59 @m4tt075 Are you guys able to identify what I am doing wrong?
Clearly the `SERVICE_COMMAND` is going wrong. I wanted to have it in `service-setup.sh` but then it keeps saying how it's not defined. 
In the `Makefile` at least it creates a package, but the variables are not put correctly into the command.

**Comment by m4tt075** [comment]
@Safihre I don't know what is right. Still learning myself. Here my advice to the best of my knowledge:
1. It seems there are several ways to define the `SERVICE_COMMAND`. It can be done in the `Makefile` directly or you go the `demoservice` route and define it in a dedicated `sabnzb.sh` file. I have taken the `.sh` file route in my `TVH` PR and think that based on all the definitions you need, this also makes sense for `sabnzb`. You can look at the `demoservice` package of my PR. Don't forget the `PLIST` addition. It is needed. 
2. Your `service-setup.sh` file looks wrong. Most of the commands you included there are already taken care of through the generic script (all the username and group creations and deletions for instance). My understanding is that `service-setup.sh` should only contain package-specific proprietary code. The rest will be auto-generated.
3. Testing-advice: You can modify the `Makefile` and `sh-files` and `src` scripts, and recompile XPE packages withouth requiring `make clean`. This is very helpful for trial and error as you can work iteratively, improving your package. I keep my VM spksrc development environment open to do the editing and `dirty` compilations, install the packages manually through the XPE Web-UI. And I keep an `ssh` session open into the XPE where I can read the package-specific `log` and `_install.log` files. Through the ssh session you can also simply execute your shell scripts to run some tests. 

Hope this helps...

**Comment by ymartin59** [comment]
I just read comment without reviewing commit - it is right. By the way, if documentation I wrote is not clear enough, please take a moment to point where it has to be improved: https://github.com/SynoCommunity/spksrc/wiki/Service-Support

**Comment by Safihre** [comment]
Aaaah, so the `SERVICE_WIZARD_GROUP` takes care of this extra `sc-download` group/user where things like SAB, NZBGet, Sonarr, Radarr etc would be in to all have access to the right files. Very nice!
@ymartin59 The `WIZARD` in the name confused me and the docs are not very clear on it's purpose. 
Maybe describe more how it is an additional user/group shared among packages?

Will try again to get this `SERVICE_COMMAND` working, it's messing with me.. Maybe I will just start from scratch with the example packages as basis.

**Comment by Safihre** [comment]
Woohoo, I finally think I am starting to understand how it all works and I got SABnzbd to run!
Will clean things up and start testing.

**Comment by Safihre** [comment]
@ymartin59 I'm stuck again. I am testing simple upgrade by uploading the same package again, which works fine for `demoservice` but for `sabnzbd` it just says "failed" and I cannot find out why.
It just won't proceed beyond `service_postuninst`. What other logs are there to inspect?
```
sabtest@XPTest:/$ cat /var/log/synopkg.log
2017/12/24 11:42:50     upgrade sabnzbd: begin to stop version 2.3.1-34
2017/12/24 11:42:50     upgrade sabnzbd 2.3.1-34 Begin start-stop-status stop
Stopping sabnzbd ...
/var/packages/sabnzbd/scripts/service-setup: line 21: type: service_poststop: not found
2017/12/24 11:42:51     upgrade sabnzbd 2.3.1-34 End start-stop-status stop ret=[0]
2017/12/24 11:42:51     upgrade sabnzbd 2.3.1-34 Begin unload apparmor
2017/12/24 11:42:51     upgrade sabnzbd 2.3.1-34 End unload apparmor ret=[0]
2017/12/24 11:42:51     upgrade sabnzbd: stop version 2.3.1-34 successfully, result 0
2017/12/24 11:42:51     upgrade sabnzbd 2.3.1-34 Begin preupgrade
/volume1/@tmp/pkginstall/scripts/service-setup: line 21: type: service_save: not found
/bin/mv: cannot stat '/*': No such file or directory
2017/12/24 11:42:51     upgrade sabnzbd 2.3.1-34 End preupgrade ret=[0]
2017/12/24 11:42:51     upgrade sabnzbd 2.3.1-34 Begin preuninst
2017/12/24 11:42:51     upgrade sabnzbd 2.3.1-34 End preuninst ret=[0]
2017/12/24 11:42:51     upgrade sabnzbd 2.3.1-34 Begin /bin/rm -rf /volume1/@appstore/sabnzbd
2017/12/24 11:42:51     upgrade sabnzbd 2.3.1-34 End /bin/rm -rf /volume1/@appstore/sabnzbd ret=[0]
2017/12/24 11:42:51     upgrade sabnzbd 2.3.1-34 Begin postuninst
2017/12/24 11:42:51     upgrade sabnzbd 2.3.1-34 End postuninst ret=[0]
2017/12/24 11:42:51     upgrade sabnzbd: Uninstall 2.3.1-34 successfully
2017/12/24 11:42:51     upgrade sabnzbd from 2.3.1-34 to 2.3.1-34 failed

sabtest@XPTest:/$ cat /tmp/sabnzbd_install.log
Sun Dec 24 11:42:51 CET 2017
Step preupgrade. USER=sabnzbd GROUP=sc-download SHARE_PATH=/volume1/downloads
Invoke service_preupgrade
service_preupgrade UPGRADE
Sun Dec 24 11:42:51 CET 2017
Step preuninst. USER=sabnzbd GROUP=sc-download SHARE_PATH=/volume1/downloads
Invoke service_preuninst
service_preuninst UPGRADE
Sun Dec 24 11:42:51 CET 2017
Step postuninst. USER=sabnzbd GROUP=sc-download SHARE_PATH=/volume1/downloads
Invoke service_postuninst
service_postuninst UPGRADE
```

**Comment by ymartin59** [comment]
Failure comes from `TO_SAVE_DIR` in `preupgrade` script... I really wonder what happened there, I have to investigate history and fix that support. https://github.com/SynoCommunity/spksrc/blob/master/mk/spksrc.service.installer#L291

**Comment by Safihre** [comment]
@ymartin59 That doesn't seem to be the problem, when I fix it by making it:
```
    $MV "${INST_VAR}" "$TMP_DIR" >> ${INST_LOG}
```
It still fails:
```
2017/12/24 12:53:41     upgrade sabnzbd: begin to stop version 2.3.1-34
2017/12/24 12:53:41     upgrade sabnzbd 2.3.1-34 Begin start-stop-status stop
Stopping sabnzbd ...
cat: /volume1/@appstore/sabnzbd/var/sabnzbd.pid: No such file or directory
/var/packages/sabnzbd/scripts/service-setup: line 21: type: service_poststop: not found
2017/12/24 12:53:41     upgrade sabnzbd 2.3.1-34 End start-stop-status stop ret=[0]
2017/12/24 12:53:41     upgrade sabnzbd 2.3.1-34 Begin unload apparmor
2017/12/24 12:53:41     upgrade sabnzbd 2.3.1-34 End unload apparmor ret=[0]
2017/12/24 12:53:41     upgrade sabnzbd: stop version 2.3.1-34 successfully, result 0
2017/12/24 12:53:41     upgrade sabnzbd 2.3.1-34 Begin preupgrade
/volume1/@tmp/pkginstall/scripts/service-setup: line 21: type: service_save: not found
2017/12/24 12:53:42     upgrade sabnzbd 2.3.1-34 End preupgrade ret=[0]
2017/12/24 12:53:42     upgrade sabnzbd 2.3.1-34 Begin preuninst
2017/12/24 12:53:42     upgrade sabnzbd 2.3.1-34 End preuninst ret=[0]
2017/12/24 12:53:42     upgrade sabnzbd 2.3.1-34 Begin /bin/rm -rf /volume1/@appstore/sabnzbd
2017/12/24 12:53:42     upgrade sabnzbd 2.3.1-34 End /bin/rm -rf /volume1/@appstore/sabnzbd ret=[0]
2017/12/24 12:53:42     upgrade sabnzbd 2.3.1-34 Begin postuninst
2017/12/24 12:53:42     upgrade sabnzbd 2.3.1-34 End postuninst ret=[0]
2017/12/24 12:53:42     upgrade sabnzbd: Uninstall 2.3.1-34 successfully
2017/12/24 12:53:42     upgrade sabnzbd from 2.3.1-34 to 2.3.1-34 failed

Sun Dec 24 12:53:41 CET 2017
Step preupgrade. USER=sabnzbd GROUP=sc-download SHARE_PATH=/volume1/downloads
Invoke service_preupgrade
service_preupgrade UPGRADE
Sun Dec 24 12:53:42 CET 2017
Step preuninst. USER=sabnzbd GROUP=sc-download SHARE_PATH=/volume1/downloads
Invoke service_preuninst
service_preuninst UPGRADE
Sun Dec 24 12:53:42 CET 2017
Step postuninst. USER=sabnzbd GROUP=sc-download SHARE_PATH=/volume1/downloads
Invoke service_postuninst
service_postuninst UPGRADE
```

**Comment by Safihre** [comment]
Fixed it!
Turns out I was doing things double in `service_postinst` and `service_postupgrade`, somehow affecting the upgrade.

**Comment by Safihre** [comment]
@ymartin59 Turns out the `TO_SAVE_DIR` thing was a bit more complicated:
Has to be cp to also copy directories.
After the copy back, the permissions are wrong.
Using just the -p option of cp would not suffice after upgrade from DSM5 or old-style package to new style.

Pushed fix to this PR!

**Comment by Safihre** [comment]
Test-building now, will remove `travis.yml` after that!

**Comment by Safihre** [comment]
[Building was success](https://travis-ci.org/Safihre/spksrc/builds/321834136) and package is tested, both new and upgrade install.
@ymartin59 What do you think? Is it good like this?

**Comment by ymartin59** [comment]
Please review my comments and discard all non-sabnzbd related commits on this branch (like python dependencies updates or compilation fixes)

**Comment by Safihre** [comment]
@ymartin59 I don't see any specifc comments, did you click `Finish review`?

I now reduced the number of commits. However, all these python dependencies are needed to even build the package on DSM6 tool-chains. `cross/python` and `native/python` don't work otherwise.
Squashing these updates even further would not be beneficial for the future, I think.

**Comment by ymartin59** [comment]
@Safihre It is difficult to review commits mixed in a single PR, and a "single commit" may be informative too if message contains enough details, it is not limited to a single line ! May you please create one PR with changes you did on "generic support framework" (two commits in #3053), probably a single commit would be enough), then a second PR with Python updates and then a third with SABnzbd update. I agree each PR depend on previous ones, but please sequence commits so that I can review and merge them easily.

File: spk/sabnzbd/src/service-setup.sh [hunk 1]
```diff
@@ -4,10 +4,35 @@ VIRTUALENV="${PYTHON_DIR}/bin/virtualenv"
...
+    fi
+fi
+
+set_all_permissions ()
+{
+    # Fix permissions for EFF_USER and GROUP on DIRNAME
+    DIRNAME=$1
+    EFF_USER=$2
+    GROUP=$3
+    echo "Granting '${EFF_USER}' user permissions on ${DIRNAME}" >> ${INST_LOG}
```
[the remark this answers is not in the document]
Comment (Safihre) [reply spk/sabnzbd/src/service-setup.sh]: Yes, you are right 👍

**Comment by Safihre** [comment]
Sure if you prefer it that way, but then it won't actually be possible to create a working `.spk` until all 3 are merged.

**Comment by ymartin59** [comment]
Yes I prefer, easier to review and double check to avoid to merge mistakes or "improvement details". Well, I can check framework with demoservice and python package can be built on its own. Thank you for this effort.

File: spk/sabnzbd/src/service-setup.sh [hunk 2]
```diff
@@ -0,0 +1,65 @@
+PYTHON_DIR="/usr/local/python"
+PATH="${SYNOPKG_PKGDEST}/bin:${SYNOPKG_PKGDEST}/env/bin:${PYTHON_DIR}/bin:${PATH}"
+VIRTUALENV="${PYTHON_DIR}/bin/virtualenv"
+PYTHON="${SYNOPKG_PKGDEST}/env/bin/python"
+SABNZBD="${SYNOPKG_PKGDEST}/share/SABnzbd/SABnzbd.py"
+CFG_FILE="${SYNOPKG_PKGDEST}/var/config.ini"
+UPRGADE_CFG_FILE="${TMP_DIR}/config.ini"
```
Comment (ymartin59) [inline spk/sabnzbd/src/service-setup.sh]: Please fix typo

**Comment by Safihre** [comment]
Submitting new ones!

**Pull request closed without merging**